src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Implementation/ImplementationEndpoint.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Implementation
+{
+    internal class ImplementationEndpoint : AbstractRazorDelegatingEndpoint<ImplementationParamsBridge, SumType<Location[]?, VSInternalReferenceItem[]?>>, IImplementationEndpoint
+    {
+        private readonly RazorDocumentMappingService _documentMappingService;
+
+        public ImplementationEndpoint(
+            DocumentContextFactory documentContextFactory,
+            LanguageServerFeatureOptions languageServerFeatureOptions,
+            RazorDocumentMappingService documentMappingService,
+            ClientNotifierServiceBase languageServer,
+            ILoggerFactory loggerFactory)
+            : base(documentContextFactory, languageServerFeatureOptions, documentMappingService, languageServer, loggerFactory.CreateLogger<ImplementationEndpoint>())
+        {
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+        }
+
+        protected override string CustomMessageTarget => RazorLanguageServerCustomMessageTargets.RazorImplementationEndpointName;
+
+        public RegistrationExtensionResult? GetRegistration(VSInternalClientCapabilities clientCapabilities)
+        {
+            const string ServerCapability = "implementationProvider";
+            var option = new ImplementationOptions();
+
+            return new RegistrationExtensionResult(ServerCapability, option);
+        }
+
+        protected override IDelegatedParams CreateDelegatedParams(ImplementationParamsBridge request, DocumentContext documentContext, Projection projection, CancellationToken cancellationToken)
+            => new DelegatedPositionParams(
+                    documentContext.Identifier,
+                    projection.Position,
+                    projection.LanguageKind);
+
+        protected async override Task<SumType<Location[]?, VSInternalReferenceItem[]?>> HandleDelegatedResponseAsync(SumType<Location[]?, VSInternalReferenceItem[]?> delegatedResponse, DocumentContext documentContext, CancellationToken cancellationToken)
```

Comment:
nit: Some of the lines in the file are pretty long, I feel like they could benefit from being broken up into multiple lines

Reply:
Yeah, I should probably introduce an alias for the sum type